Dockerfile

```diff
@@ -1,36 +1,68 @@
-FROM docker:18.06.1 as docker
-FROM python:3.7.2-stretch
+ARG DOCKER_VERSION=18.09.5
+ARG PYTHON_VERSION=3.7.3
+ARG ALPINE_VERSION=3.9
+ARG DEBIAN_VERSION=stretch-slim
+ARG BUILD_PLATFORM=alpine
 
-RUN set -ex; \
-    apt-get update -qq; \
-    apt-get install -y \
-        locales \
-        python-dev \
-        git
+FROM docker:${DOCKER_VERSION} AS docker-cli
 
-COPY --from=docker /usr/local/bin/docker /usr/local/bin/docker
+FROM python:${PYTHON_VERSION}-alpine${ALPINE_VERSION} AS build-alpine
+RUN apk add --no-cache \
+    bash \
+    build-base \
+    ca-certificates \
+    curl \
+    gcc \
+    git \
+    libc-dev \
+    libffi-dev \
+    libgcc \
+    make \
+    musl-dev \
+    openssl \
+    openssl-dev \
+    python2 \
+    python2-dev \
+    zlib-dev
+ENV BUILD_BOOTLOADER=1
 
-# Python3 requires a valid locale
-RUN echo "en_US.UTF-8 UTF-8" > /etc/locale.gen && locale-gen
-ENV LANG en_US.UTF-8
+FROM python:${PYTHON_VERSION}-slim-stretch AS build-debian
+RUN apt-get update && apt-get install -y \
+    curl \
+    gcc \
+    git \
+    libc-dev \
+    libgcc-6-dev \
+    make \
+    openssl \
+    python2.7-dev
 
-RUN useradd -d /home/user -m -s /bin/bash user
+FROM build-${BUILD_PLATFORM} AS build
 WORKDIR /code/
-
 # FIXME(chris-crone): virtualenv 16.3.0 breaks build, force 16.2.0 until fixed
-RUN pip install virtualenv==16.2.0
-RUN pip install tox==2.9.1
+RUN pip install virtualenv==16.2.0 tox==2.1.1
 
-ADD requirements.txt /code/
-ADD requirements-dev.txt /code/
-ADD .pre-commit-config.yaml /code/
-ADD setup.py /code/
-ADD tox.ini /code/
-ADD compose /code/compose/
-ADD README.md /code/
+ARG GIT_COMMIT=unknown
+ENV DOCKER_COMPOSE_GITSHA=$GIT_COMMIT
+COPY requirements.txt \
```

Comment:
Same as for the `pip install`; keeping these separate may actually be better (but give it a try), so that caching is optimized. The files don't overlap, so combining `COPY` lines won't make the image smaller. It would reduce the number of layers, but this is an intermediate stage, so won't affect the final image

Reply:
Done